{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "from matplotlib import pyplot as plt\n",
    "\n",
    "import json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def call_RNAfold(sequence):\n",
    "    MyOut = subprocess.Popen(['RNAfold', '-p', '--noPS', '--constraint'],\n",
    "            stdin=subprocess.PIPE,\n",
    "            stdout=subprocess.PIPE, \n",
    "            stderr=subprocess.STDOUT)\n",
    "    stdout, stderr = MyOut.communicate(input=str.encode(sequence))\n",
    "    return stdout\n",
    "\n",
    "def get_energy_RNAfold(stdout_string):\n",
    "    temp = stdout_string.decode(\"utf-8\") \n",
    "    energy_line = temp.split('\\n')[-5]\n",
    "    energy_val = energy_line[energy_line.index(' '):]\n",
    "    energy_val = energy_val.strip().strip('()').strip()\n",
    "    \n",
    "    mfe_line = temp.split('\\n')[-4]\n",
    "    mfe_val = mfe_line[energy_line.index(' '):]\n",
    "    mfe_val = mfe_val.strip().strip('[]').strip()\n",
    "    return float(energy_val), float(mfe_val)    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Establish some basic dataframes that should contain all the information we need"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "\n",
    "genome_location = './ecoli.gb'\n",
    "genome_type = 'genbank'\n",
    "###Read genome\n",
    "genome = list(SeqIO.parse(genome_location, genome_type))\n",
    "\n",
    "###Good to double check. If this fails on a future genome we'll have to decide what to do\n",
    "assert len(genome) == 1\n",
    "genome = genome[0]\n",
    "\n",
    "###Set up empty lists for now\n",
    "###These are for the gene dataframe\n",
    "loci = []\n",
    "starts = []\n",
    "stops = []\n",
    "strands = []\n",
    "###These are for the start codon dataframe\n",
    "loci_long = []\n",
    "ATG_indices = []\n",
    "\n",
    "\n",
    "###And iterate through the genome\n",
    "for feature in genome.features:\n",
    "    if feature.type == 'CDS':\n",
    "        ###Filter out the weirdos\n",
    "        if 'pseudo' in feature.qualifiers or 'ribosomal_slippage' in feature.qualifiers:\n",
    "            continue\n",
    "        ###Make sure that no locus_tag appears twice in the genome\n",
    "        ###If this happens on some future genome then we'll have to figure out what to do about it\n",
    "        locus_tag = feature.qualifiers['locus_tag'][0]\n",
    "        assert locus_tag not in loci\n",
    "        ###Extract the coding sequence\n",
    "        seq = str(feature.extract(genome.seq))\n",
    "        ###Double check that all coding sequences are a multiple of 3\n",
    "        assert len(seq)%3 == 0\n",
    "        assert feature.location.end - feature.location.start == len(seq)\n",
    "        ###Append some basic data\n",
    "        loci.append(locus_tag)\n",
    "        starts.append(feature.location.start)\n",
    "        stops.append(feature.location.end)\n",
    "        strands.append(feature.location.strand)\n",
    "        \n",
    "        ###Now on to ATGS\n",
    "        ###This is a quick/efficient way to find the occurrence of all substrings within a string\n",
    "        ATG_locs = [m.start() for m in re.finditer('ATG', seq)]\n",
    "        ###Iterate through all ATG indices and append the data\n",
    "        for loc in ATG_locs:\n",
    "            loci_long.append(locus_tag)\n",
    "            ATG_indices.append(loc)\n",
    "\n",
    "###Create and clean up a gene level dataframe\n",
    "col_names = ['start', 'stop', 'strand']\n",
    "gene_df = pd.DataFrame(list(zip(starts, stops, strands)), columns=col_names, index=loci)\n",
    "\n",
    "###Create and clean up a start codon data frame\n",
    "###Now make this all into a nice and easy pandas dataframe\n",
    "col_names = ['locus_tag', 'ATG_index']\n",
    "start_df = pd.DataFrame(list(zip(loci_long, ATG_indices)), columns=col_names)\n",
    "\n",
    "###These last two columns we might want are kind of unnecessary/re-dundant but here is a quick way\n",
    "###to calculate them\n",
    "###First set the default as Internal\n",
    "start_df['location'] = 'Internal'\n",
    "###And make it External for every ATG_index of zero\n",
    "start_df.at[start_df[start_df['ATG_index'] == 0].index, 'location'] = 'External'\n",
    "\n",
    "###Same thing here. Set the default as False\n",
    "start_df['within_frame'] = False\n",
    "###And now say if the index divided by 3 has no remainder, set it to True\n",
    "start_df.at[start_df[start_df['ATG_index']%3 == 0].index, 'within_frame'] = True\n",
    "###And if we want the External ATGs to not have a within_frame value we can now just say\n",
    "start_df.at[start_df[start_df['location'] == 'External'].index, 'within_frame'] = ''\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "###Writing pandas dataframes are crazy easy and so much easier/more common than csvwrite/reader\n",
    "###Within the sep argument, you can make it comma separated, tab separated, etc.\n",
    "###I usually prefer tab separated files (tsv) for no particular reason.\n",
    "###The index=False statement can be True or False. It really only affects the way you read the files\n",
    "###later on\n",
    "\n",
    "gene_df.to_csv('../Data/ecoli_gene_df.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Let's look at those dataframes a bit"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>locus_tag</th>\n",
       "      <th>ATG_index</th>\n",
       "      <th>location</th>\n",
       "      <th>within_frame</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>b0001</td>\n",
       "      <td>0</td>\n",
       "      <td>External</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>b0002</td>\n",
       "      <td>0</td>\n",
       "      <td>External</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>b0002</td>\n",
       "      <td>37</td>\n",
       "      <td>Internal</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>b0002</td>\n",
       "      <td>79</td>\n",
       "      <td>Internal</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>b0002</td>\n",
       "      <td>147</td>\n",
       "      <td>Internal</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  locus_tag  ATG_index  location within_frame\n",
       "0     b0001          0  External             \n",
       "1     b0002          0  External             \n",
       "2     b0002         37  Internal        False\n",
       "3     b0002         79  Internal        False\n",
       "4     b0002        147  Internal         True"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "start_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(77344, 4)\n"
     ]
    }
   ],
   "source": [
    "print(start_df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>b0001</th>\n",
       "      <td>189</td>\n",
       "      <td>255</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0002</th>\n",
       "      <td>336</td>\n",
       "      <td>2799</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0003</th>\n",
       "      <td>2800</td>\n",
       "      <td>3733</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0004</th>\n",
       "      <td>3733</td>\n",
       "      <td>5020</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0005</th>\n",
       "      <td>5233</td>\n",
       "      <td>5530</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       start  stop  strand\n",
       "b0001    189   255       1\n",
       "b0002    336  2799       1\n",
       "b0003   2800  3733       1\n",
       "b0004   3733  5020       1\n",
       "b0005   5233  5530       1"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(4239, 3)\n"
     ]
    }
   ],
   "source": [
    "print(gene_df.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Now calculate all of the relevant folding energies and add them to the dataframe\n",
    "\n",
    "As (currently) written, this code should extract 50 nt's before the start codon, the start codon itself, and then 47 nt's downstream of the start codon (for a total of 100 nt's). This decision is entirely arbitrary and the source of much contention/error but no one knows a better way to do it.\n",
    "\n",
    "From that 100 nt fragment, we can calculate the folding energy (both the minimum folding energy or the folding energy of the thermodynamic ensemble). We can also add in a constraint that some region has to be unpaired (like the SD region) and ask how the folding energy changes between the unconstrained and constrained values. \n",
    "\n",
    "Note: On my laptop, for the E. coli genome with ~80,000 ATGs, this code took about 2 hours to run. So not bad all things considered but will get annoying as we consider GTG, TTG, etc. codons one day."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "flanking_bases = 50\n",
    "seq_len = 100\n",
    "\n",
    "###Important to check this later. These are the basepair constraints used for calculating\n",
    "###free energy under the condition that we constrain the SD region to be open\n",
    "constraint = ('.'*34) + ('x'*12)+ ('.'*54)\n",
    "assert len(constraint) == seq_len\n",
    "\n",
    "###Iterate through all the start codons\n",
    "for index in start_df.index[:]:\n",
    "    ###Get this info directly from the index\n",
    "    locus = start_df.at[index, 'locus_tag']\n",
    "    start_index = start_df.at[index, 'ATG_index']\n",
    "    ###And this info from the gene_df\n",
    "    strand = gene_df.at[locus, 'strand']\n",
    "    ###Get the sequence (strand dependent)\n",
    "    if strand == 1:\n",
    "        start = gene_df.at[locus, 'start']\n",
    "        ###Ensure that I have found an ATG\n",
    "        if genome.seq[start+start_index:start+start_index+3] != 'ATG':\n",
    "            print('Something is rotten in the state of denmark')\n",
    "            continue\n",
    "        seq = genome.seq[start+start_index-flanking_bases:start+start_index+flanking_bases]\n",
    "\n",
    "    elif strand == -1:\n",
    "        start = gene_df.at[locus, 'stop']\n",
    "        if genome.seq[start-start_index-3:start-start_index].reverse_complement() != 'ATG':\n",
    "            print('Something is rotten in the state of denmark, on the other side')\n",
    "            continue    \n",
    "        seq = genome.seq[start-start_index-flanking_bases:start-start_index+flanking_bases].reverse_complement()\n",
    "    \n",
    "    ###Just like quadruple checking that this is true\n",
    "    assert len(seq) == seq_len\n",
    "    \n",
    "    ###Call RNAfold on the sequence!\n",
    "    RNA_out = call_RNAfold(str(seq))\n",
    "    a, b = get_energy_RNAfold(RNA_out)\n",
    "    start_df.at[index, 'Unconstrained_mfe'] = b\n",
    "    start_df.at[index, 'Unconstrained_ensemble'] = a\n",
    "\n",
    "    ###Now call it in the constrained form such that the SD region is unpaired\n",
    "    RNA_out = call_RNAfold(str(seq)+'\\n'+constraint)\n",
    "    a, b = get_energy_RNAfold(RNA_out)\n",
    "    start_df.at[index, 'Constrained_mfe'] = b\n",
    "    start_df.at[index, 'Constrained_ensemble'] = a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>locus_tag</th>\n",
       "      <th>ATG_index</th>\n",
       "      <th>location</th>\n",
       "      <th>within_frame</th>\n",
       "      <th>Unconstrained_mfe</th>\n",
       "      <th>Unconstrained_ensemble</th>\n",
       "      <th>Constrained_mfe</th>\n",
       "      <th>Constrained_ensemble</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>b0001</td>\n",
       "      <td>0</td>\n",
       "      <td>External</td>\n",
       "      <td></td>\n",
       "      <td>-4.30</td>\n",
       "      <td>-2.0</td>\n",
       "      <td>-3.47</td>\n",
       "      <td>-1.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>b0002</td>\n",
       "      <td>0</td>\n",
       "      <td>External</td>\n",
       "      <td></td>\n",
       "      <td>-20.68</td>\n",
       "      <td>-18.1</td>\n",
       "      <td>-17.03</td>\n",
       "      <td>-14.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>b0002</td>\n",
       "      <td>37</td>\n",
       "      <td>Internal</td>\n",
       "      <td>False</td>\n",
       "      <td>-29.91</td>\n",
       "      <td>-27.0</td>\n",
       "      <td>-24.78</td>\n",
       "      <td>-22.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>b0002</td>\n",
       "      <td>79</td>\n",
       "      <td>Internal</td>\n",
       "      <td>False</td>\n",
       "      <td>-43.35</td>\n",
       "      <td>-40.8</td>\n",
       "      <td>-41.45</td>\n",
       "      <td>-39.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>b0002</td>\n",
       "      <td>147</td>\n",
       "      <td>Internal</td>\n",
       "      <td>True</td>\n",
       "      <td>-22.77</td>\n",
       "      <td>-20.5</td>\n",
       "      <td>-17.23</td>\n",
       "      <td>-15.8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  locus_tag  ATG_index  location within_frame  Unconstrained_mfe  \\\n",
       "0     b0001          0  External                           -4.30   \n",
       "1     b0002          0  External                          -20.68   \n",
       "2     b0002         37  Internal        False             -29.91   \n",
       "3     b0002         79  Internal        False             -43.35   \n",
       "4     b0002        147  Internal         True             -22.77   \n",
       "\n",
       "   Unconstrained_ensemble  Constrained_mfe  Constrained_ensemble  \n",
       "0                    -2.0            -3.47                  -1.4  \n",
       "1                   -18.1           -17.03                 -14.5  \n",
       "2                   -27.0           -24.78                 -22.7  \n",
       "3                   -40.8           -41.45                 -39.8  \n",
       "4                   -20.5           -17.23                 -15.8  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "start_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>locus_tag</th>\n",
       "      <th>ATG_index</th>\n",
       "      <th>location</th>\n",
       "      <th>within_frame</th>\n",
       "      <th>Unconstrained_mfe</th>\n",
       "      <th>Unconstrained_ensemble</th>\n",
       "      <th>Constrained_mfe</th>\n",
       "      <th>Constrained_ensemble</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>77339</th>\n",
       "      <td>b4403</td>\n",
       "      <td>496</td>\n",
       "      <td>Internal</td>\n",
       "      <td>False</td>\n",
       "      <td>-19.76</td>\n",
       "      <td>-17.8</td>\n",
       "      <td>-16.16</td>\n",
       "      <td>-15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77340</th>\n",
       "      <td>b4403</td>\n",
       "      <td>543</td>\n",
       "      <td>Internal</td>\n",
       "      <td>True</td>\n",
       "      <td>-22.52</td>\n",
       "      <td>-19.9</td>\n",
       "      <td>-18.91</td>\n",
       "      <td>-16.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77341</th>\n",
       "      <td>b4403</td>\n",
       "      <td>574</td>\n",
       "      <td>Internal</td>\n",
       "      <td>False</td>\n",
       "      <td>-23.84</td>\n",
       "      <td>-21.0</td>\n",
       "      <td>-16.86</td>\n",
       "      <td>-15.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77342</th>\n",
       "      <td>b4403</td>\n",
       "      <td>639</td>\n",
       "      <td>Internal</td>\n",
       "      <td>True</td>\n",
       "      <td>-22.39</td>\n",
       "      <td>-20.9</td>\n",
       "      <td>-17.05</td>\n",
       "      <td>-15.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77343</th>\n",
       "      <td>b4403</td>\n",
       "      <td>658</td>\n",
       "      <td>Internal</td>\n",
       "      <td>False</td>\n",
       "      <td>-23.35</td>\n",
       "      <td>-20.8</td>\n",
       "      <td>-15.88</td>\n",
       "      <td>-14.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      locus_tag  ATG_index  location within_frame  Unconstrained_mfe  \\\n",
       "77339     b4403        496  Internal        False             -19.76   \n",
       "77340     b4403        543  Internal         True             -22.52   \n",
       "77341     b4403        574  Internal        False             -23.84   \n",
       "77342     b4403        639  Internal         True             -22.39   \n",
       "77343     b4403        658  Internal        False             -23.35   \n",
       "\n",
       "       Unconstrained_ensemble  Constrained_mfe  Constrained_ensemble  \n",
       "77339                   -17.8           -16.16                 -15.0  \n",
       "77340                   -19.9           -18.91                 -16.9  \n",
       "77341                   -21.0           -16.86                 -15.1  \n",
       "77342                   -20.9           -17.05                 -15.8  \n",
       "77343                   -20.8           -15.88                 -14.1  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "start_df.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "###Write this data to a file\n",
    "start_df.to_csv('../Data/ecoli_folding_energies.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  },
  "toc": {
   "colors": {
    "hover_highlight": "#DAA520",
    "navigate_num": "#000000",
    "navigate_text": "#333333",
    "running_highlight": "#FF0000",
    "selected_highlight": "#FFD700",
    "sidebar_border": "#EEEEEE",
    "wrapper_background": "#FFFFFF"
   },
   "moveMenuLeft": true,
   "nav_menu": {
    "height": "66px",
    "width": "252px"
   },
   "navigate_menu": true,
   "number_sections": true,
   "sideBar": true,
   "threshold": 4,
   "toc_cell": false,
   "toc_section_display": "block",
   "toc_window_display": false,
   "widenNotebook": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
